### PR TITLE
rustc: Add a new crate type, cdylib

### LINKED ISSUE #33132: Tracking issue for adding a new cdylib crate type
Tracking issue for https://github.com/rust-lang/rfcs/pull/1510

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an implementation of [RFC 1510](https://github.com/rust-lang/rfcs/pull/1510) which adds a new crate type,
`cdylib`, to the compiler. This new crate type differs from the existing `dylib`
crate type in a few key ways:
- No metadata is present in the final artifact
- Symbol visibility rules are the same as executables, that is only reachable
  `extern` functions are visible symbols
- LTO is allowed
- All libraries are always linked statically

This commit is relatively simple by just plubming the compiler with another
crate type which takes different branches here and there. The only major change
is an implementation of the `Linker::export_symbols` function on Unix which now
actually does something. This helps restrict the public symbols from a cdylib on
Unix.

With this PR a "hello world" `cdylib` is 7.2K while the same `dylib` is 2.4MB,
which is some nice size savings!

Closes #33132

### CODE REVIEW

File: src/librustc_trans/back/linker.rs [hunk 1]
```diff
@@ -198,8 +201,30 @@ impl<'a> Linker for GnuLinker<'a> {
...
+    fn export_symbols(&mut self,
+                      sess: &Session,
+                      trans: &CrateTranslation,
+                      tmpdir: &Path,
+                      crate_type: CrateType) {
+        let path = tmpdir.join("list");
+        let res = (|| -> io::Result<()> {
+            let mut f = BufWriter::new(File::create(&path)?);
+            for sym in exported_symbols(sess, trans, crate_type) {
+                writeln!(f, "  {}", sym)?;
```
Comment (eddyb) [inline src/librustc_trans/back/linker.rs]: This doesn't seem to be enough for iOS, [see failure](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/279/steps/compile/logs/stdio).

**Comment by alexcrichton** [comment]
Indeed! looks quite suspicious

**Comment by Manishearth** [comment]
Travis fails

**Comment by eddyb** [comment]
I'm working on rebasing this over #33602 because otherwise I'd need windows to investigate my metadata problem.
Could be delayed for a few days by my flight, so if you don't hear anything from me, feel free to merge this.

**Comment by alexcrichton** [comment]
@bors: r=brson 557aae0

@eddyb I've applied your patch to retain the metadata symbol for dylibs, let's see how far it gets...

File: src/test/run-make/cdylib/Makefile [hunk 2]
```diff
@@ -0,0 +1,10 @@
+include ../tools.mk
+
+all:
+	$(RUSTC) bar.rs
+	$(RUSTC) foo.rs
+	$(CC) $(CFLAGS) foo.c -lfoo -o $(call RUN_BINFILE,foo) -L $(TMPDIR)
```
Comment (retep998) [inline src/test/run-make/cdylib/Makefile]: This won't work for msvc.

**Comment by bluss** [comment]
This is kind of immediately a new feature on the stable release where this is included, isn't it? If I'm wrong, then the relnote tag can be removed again.

**Comment by torkleyy** [comment]
@alexcrichton How can I compile a crate as `cdylib` without modifying `Cargo.toml`? I thought it would be `cargo rustc -- --crate-type cdylib`, but the `crate-type` option is just ignored.

**Comment by alexcrichton** [comment]
@torkleyy unfortunately that's not possible right now (short of generating a new `Cargo.toml`. Would you like to open a new issue at rust-lang/cargo?

**Comment by torkleyy** [comment]
@alexcrichton Thanks for your reply!

> Would you like to open a new issue at rust-lang/cargo?

Sure, I'll do that.

**Pull request merged**